File: NetFramework/TreeView/VB/BoundModeITypedList/Employees/TypedList.vb
Imports System.ComponentModel

<Serializable>
Public Class TypedList(Of T)
    Inherits BindingList(Of T)
    Implements ITypedList

    <NonSerialized>
    Private properties As PropertyDescriptorCollection

    Public Sub New()
        MyBase.New()
        Dim pdc As PropertyDescriptorCollection = TypeDescriptor.GetProperties(GetType(T), New Attribute() {New BrowsableAttribute(True)})
        properties = pdc
    End Sub

#Region "ITypedList Implementation"

    Private Function ITypedList_GetItemProperties(listAccessors() As PropertyDescriptor) As PropertyDescriptorCollection Implements ITypedList.GetItemProperties
        Dim pdc As PropertyDescriptorCollection

        If listAccessors IsNot Nothing AndAlso listAccessors.Length > 0 Then
            pdc = ListBindingHelper.GetListItemProperties(listAccessors(0).PropertyType)
        Else
            pdc = properties
        End If

        Return pdc
    End Function

    Private Function ITypedList_GetListName(listAccessors() As PropertyDescriptor) As String Implements ITypedList.GetListName
        Return GetType(T).Name
    End Function

#End Region
End Class

File: NetFramework/TreeView/CS/BoundModeITypedList/Employees/TypedList.cs
using System;
using System.ComponentModel;
using System.Windows.Forms;

namespace SamplesData
{
    [Serializable()]
    public class TypedList<T> : BindingList<T>, ITypedList
    {
        [NonSerialized()]
        private PropertyDescriptorCollection properties;

        public TypedList() : base()
        {
            PropertyDescriptorCollection pdc = TypeDescriptor.GetProperties(
                typeof(T),
                new Attribute[] { new BrowsableAttribute(true) });

            properties = pdc;
        }

        #region ITypedList Implementation

        public PropertyDescriptorCollection GetItemProperties(PropertyDescriptor[] listAccessors)
        {
            PropertyDescriptorCollection pdc;

            if (listAccessors != null && listAccessors.Length > 0)
            {
                pdc = ListBindingHelper.GetListItemProperties(listAccessors[0].PropertyType);
            }
            else
            {
                pdc = properties;
            }

            return pdc;
        }

        public string GetListName(PropertyDescriptor[] listAccessors)
        {
            return typeof(T).Name;
        }

        #endregion
    }
}
